src/xdist/dsession.py

```diff
@@ -182,6 +182,15 @@ def worker_workerfinished(self, node):
             self.shouldstop = f"{node} received keyboard-interrupt"
             self.worker_errordown(node, "keyboard-interrupt")
             return
+
+        shouldfail = node.workeroutput["shouldfail"]
+        shouldstop = node.workeroutput["shouldstop"]
+        for shouldx in [shouldfail, shouldstop]:
+            if shouldx:
+                self.shouldstop = shouldx
```

Comment:
If `shouldstop` is already set (which I think is possible if raced with another worker?), I think we shouldn't overwrite it.